skimage/restoration/j_invariant.py

```diff
@@ -0,0 +1,309 @@
+import itertools
+import functools
+
+import numpy as np
+from scipy import ndimage as ndi
+
+from ..measure import compare_mse
+from ..util import img_as_float
+
+
+def _interpolate_image(image, multichannel=False):
+    """Replacing each pixel in ``image`` with the average of its neighbors.
+
+    Parameters
+    ----------
+    image : ndarray
+        Input data to be interpolated.
+    multichannel : bool, optional
+        Whether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.
+
+    Returns
+    -------
+    interp : ndarray
+        Interpolated version of `image`.
+    """
+    spatialdims = image.ndim if not multichannel else image.ndim - 1
+    conv_filter = ndi.generate_binary_structure(spatialdims, 1).astype(float)
+    conv_filter.ravel()[conv_filter.size // 2] = 0
+    conv_filter /= conv_filter.sum()
+
+    if multichannel:
+        interp = np.zeros(image.shape)
+        for i in range(image.shape[-1]):
+            interp[..., i] = ndi.convolve(image[..., i], conv_filter,
+                                          mode='mirror')
+    else:
+        interp = ndi.convolve(image, conv_filter, mode='mirror')
+    return interp
+
+
+def _generate_grid_slice(shape, *, offset, stride=3):
+    """Generate slices of uniformly-spaced points in an array.
+
+    Parameters
+    ----------
+    shape : tuple of int
+        Shape of the mask.
+    offset : int
+        The offset of the grid of ones. Iterating over ``offset`` will cover
+        the entire array. It should be between 0 and ``stride ** ndim``, not
+        inclusive, where ``ndim = len(shape)``.
+    stride : int, optional
+        The spacing between ones, used in each dimension.
+
+    Returns
+    -------
+    mask : ndarray
+        The mask.
+
+    Examples
+    --------
+    >>> shape = (4, 4)
+    >>> array = np.zeros(shape, dtype=int)
+    >>> grid_slice = _generate_grid_slice(shape, offset=0, stride=2)
+    >>> array[grid_slice] = 1
+    >>> print(array)
+    [[1 0 1 0]
+     [0 0 0 0]
+     [1 0 1 0]
+     [0 0 0 0]]
+
+    Changing the offset moves the location of the 1s:
+
+    >>> array = np.zeros(shape, dtype=int)
+    >>> grid_slice = _generate_grid_slice(shape, offset=3, stride=2)
+    >>> array[grid_slice] = 1
+    >>> print(array)
+    [[0 0 0 0]
+     [0 1 0 1]
+     [0 0 0 0]
+     [0 1 0 1]]
+    """
+    phases = np.unravel_index(offset, (stride,) * len(shape))
+    mask = tuple(slice(p, None, stride) for p in phases)
+
+    return mask
+
+
+def invariant_denoise(image, denoise_function, *, stride=4, multichannel=False,
```

Comment:
with the recent merge of #4446 all denoising functions in scikit-image now take a `multichannel` argument so maybe we can rely on `denoiser_kwargs` instead?